lib/ansible/modules/system/gconftool2.py

```diff
@@ -197,74 +285,28 @@ def main():
-        if direct and config_source is None:
-            module.fail_json(msg='If "direct" is "yes" then the ' +
-                             '"config_source" must be specified')
-        elif not direct and config_source is not None:
-            module.fail_json(msg='If the "config_source" is specified ' +
-                             'then "direct" must be "yes"')
-
-    # Create a gconf2 preference
-    gconf_pref = GConf2Preference(module, key, value_type,
-                                  value, direct, config_source)
-    # Now we get the current value, if not found don't fail
-    _, current_value = gconf_pref.call("get", fail_onerr=False)
-
-    # Check if the current value equals the value we want to set.  If not, make
-    # a change
-    if current_value != value:
-        # If check mode, we know a change would have occured.
-        if module.check_mode:
-            # So we will set the change to True
-            change = True
-            # And set the new_value to the value that would have been set
-            new_value = value
-        # If not check mode make the change.
-        else:
-            change, new_value = gconf_pref.call(state)
-    # If the value we want to set is the same as the current_value, we will
-    # set the new_value to the current_value for reporting
-    else:
-        new_value = current_value
-
-    facts = dict(gconftool2={'changed': change,
-                             'key': key,
-                             'value_type': value_type,
-                             'new_value': new_value,
-                             'previous_value': current_value,
-                             'playbook_value': module.params['value']})
-
-    module.exit_json(changed=change, ansible_facts=facts)
+    # If direct access was requested, config_source must be provided.
+    if module.params['direct'] and (module.params['config_source'] is None):
+        module.fail_json(msg='If "direct" is "yes", then the "config_source" must be specified')
+
+    # If present state was specified, value and value type must be provided.
+    if module.params['state'] == 'present' and (module.params['value'] is None or module.params['value_type'] is None):
+        module.fail_json(msg='State "present" requires both "value" and "value_type" to be set.')
+
+    # Create wrapper instance.
+    gconftool2 = GConfTool2(module, module.params['direct'], module.params['config_source'], module.check_mode)
+
+    # Process based on different states.
+    if module.params['state'] == 'get':
+        value = gconftool2.get(module.params['key'])
+        module.exit_json(changed=False, value=value)
+    elif module.params['state'] == 'present':
+        changed = gconftool2.set(module.params['key'], module.params['value_type'], module.params['value'])
+        module.exit_json(changed=changed)
+    elif module.params['state'] == 'absent':
+        changed = gconftool2.unset(module.params['key'])
+        module.exit_json(changed=changed)
```

Comment:
Ok, I can confirm my rationale for breaking backwards compatibility above. When using `ansible_facts=facts`, you end-up merging the values into Ansible facts, which I consider wrong in respect with this module - since it's not one of the generic `*_facts` modules. The value changes on each invocation, overwriting what was previously set, it gives limited view of what values are available (just one, the last one changed/read etc), whereas the `*_facts` modules are usually (I think?) invoked once, and they provide a complete picture of facts they cover.

E.g., if you invoke `xattr` module, you need to explicitly `register` the variable in a task, and it is close behaviour-wise to `gconftool2`.

I do not see why `gconftool2` should receive special treatment in that regard, especially since it does not behave as a kind of "facts" module (like `setup` and similar).

Although, I do store less information in my refactoring as well - so that's another thing to discuss after we settle on backwards compatibility decision.